### PR TITLE
Support for DDR4 DDIMM parsing

### PR DESCRIPTION
**Pull request by branupama**
This commit adds the code to support parsing of DDR4 DDIMM type VPD.

ParserFactory is updated to handle the DDR4 type to get the DDIMM VPD
parsing instance for DDR4 DDIMM and DdimmVpdParser class updated to
handle DDR4 DDIMM VPD parsing with new API to calculate the
DDR4 DDIMM size.

Tested the changes by running vpd-manager application, observed DDIMM
parser is selected for DDR4 DDIMM EEPROM files and parsing is
successful. And also observed, DDIMM size is calculated correctly
for DDR4 DDIMMs.

### CODE REVIEW

File: src/ddimm_parser.cpp [hunk 1]
```diff
@@ -209,11 +297,16 @@ size_t
     {
         l_dimmSize = getDdr5BasedDdimmSize(i_iterator);
     }
+    else if ((i_iterator[constants::SPD_BYTE_2] & constants::SPD_BYTE_MASK) ==
```
Comment (jinuthomas) [inline src/ddimm_parser.cpp:300]: why do you need an  & here ?

File: src/ddimm_parser.cpp [hunk 2]
```diff
@@ -200,6 +200,94 @@ size_t DdimmVpdParser::getDdr5BasedDdimmSize(
     return constants::CONVERT_GB_TO_KB * l_dimmSize;
 }
 
+size_t DdimmVpdParser::getDdr4BasedDdimmSize(
+    types::BinaryVector::const_iterator i_iterator)
+{
+    size_t l_tmpValue = 0, l_dimmSize = 0;
```
Comment (jinuthomas) [inline src/ddimm_parser.cpp:206]: why is l_tmpValue a size_t ?

File: src/ddimm_parser.cpp [hunk 3]
```diff
@@ -200,6 +200,94 @@ size_t DdimmVpdParser::getDdr5BasedDdimmSize(
...
+
+    if (l_tmpValue > constants::JEDEC_RESERVED_BITS)
+    {
+        logging::logMessage(
+            "Bad data in vpd byte 13. Can't calculate primary bus width "
+            "and so dimm size.");
+        return l_dimmSize;
+    }
+
+    size_t l_primaryBusWid = 1;
```
Comment (jinuthomas) [inline src/ddimm_parser.cpp:237]: why size_t ?

File: src/ddimm_parser.cpp [hunk 4]
```diff
@@ -200,6 +200,94 @@ size_t DdimmVpdParser::getDdr5BasedDdimmSize(
...
+
+    if (l_tmpValue > constants::JEDEC_RESERVED_BITS)
+    {
+        logging::logMessage(
+            "Bad data in vpd byte 12. Can't calculate SDRAM width and so "
+            "dimm size.");
+        return l_dimmSize;
+    }
+
+    size_t l_sdramWidth = 1;
```
Comment (jinuthomas) [inline src/ddimm_parser.cpp:253]: why size_t ?

File: src/ddimm_parser.cpp [hunk 5]
```diff
@@ -200,6 +200,94 @@ size_t DdimmVpdParser::getDdr5BasedDdimmSize(
...
+    l_tmpValue >>= constants::JEDEC_RESERVED_BITS;
+
+    if (l_tmpValue > constants::JEDEC_RESERVED_BITS)
+    {
+        logging::logMessage(
+            "Can't calculate number of ranks. Invalid data found.");
+        return l_dimmSize;
+    }
+
+    size_t l_logicalRanksPerDimm = (l_tmpValue + 1) * l_dieCount;
```
Comment (jinuthomas) [inline src/ddimm_parser.cpp:283]: why size_t ?

File: src/ddimm_parser.cpp [hunk 6]
```diff
@@ -200,6 +200,94 @@ size_t DdimmVpdParser::getDdr5BasedDdimmSize(
...
+    }
+
+    /* Calculate Number of ranks */
+    l_tmpValue = i_iterator[constants::SPD_BYTE_12] &
+                 constants::JEDEC_NUM_RANKS_MASK;
+    l_tmpValue >>= constants::JEDEC_RESERVED_BITS;
+
+    if (l_tmpValue > constants::JEDEC_RESERVED_BITS)
+    {
+        logging::logMessage(
```
Comment (jinuthomas) [inline src/ddimm_parser.cpp:278]: TODO:: need an error log here

**Review by jinuthomas — CHANGES_REQUESTED** [review]

File: src/ddimm_parser.cpp (hunk 2 above)
Comment (branupama) [reply src/ddimm_parser.cpp]: Updated

File: src/ddimm_parser.cpp (hunk 3 above)
Comment (branupama) [reply src/ddimm_parser.cpp]: updated

File: src/ddimm_parser.cpp (hunk 4 above)
Comment (branupama) [reply src/ddimm_parser.cpp]: updated

File: src/ddimm_parser.cpp (hunk 5 above)
Comment (branupama) [reply src/ddimm_parser.cpp]: updated

File: src/ddimm_parser.cpp (hunk 6 above)
Comment (branupama) [reply src/ddimm_parser.cpp]: Done

File: src/ddimm_parser.cpp (hunk 1 above)
Comment (branupama) [reply src/ddimm_parser.cpp]: removed

**Review by jinuthomas — APPROVED** [review]

**Pull request merged**